services/circleci/circleci.tester.js

```diff
@@ -9,62 +9,58 @@ const t = (module.exports = new ServiceTester({
 }))
 
 t.create('circle ci (valid, without branch)')
-  .get('/project/github/RedSparr0w/node-csgo-parser.json')
+  .get('/build/gh/RedSparr0w/node-csgo-parser.json')
   .expectBadge({
     label: 'build',
     message: isBuildStatus,
   })
 
-t.create('circle ci (valid, with branch)')
-  .get('/project/github/RedSparr0w/node-csgo-parser/master.json')
+t.create('circle ci (valid repo, valid branch)')
+  .get('/build/gh/RedSparr0w/node-csgo-parser/master.json')
   .expectBadge({
     label: 'build',
     message: isBuildStatus,
   })
 
-t.create('build status with "github" as a default VCS')
-  .get('/project/RedSparr0w/node-csgo-parser/master.json')
+t.create('circle ci (valid repo, invalid branch)')
+  .get('/build/gh/RedSparr0w/node-csgo-parser/not-a-real-branch.json')
   .expectBadge({
     label: 'build',
-    message: isBuildStatus,
+    message: 'no builds',
   })
 
+t.create('circle ci (not found)')
+  .get('/build/gh/PyvesB/EmptyRepo.json')
+  .expectBadge({ label: 'build', message: 'project not found' })
+
 t.create('circle ci (valid, with token)')
   .get(
-    '/token/b90b5c49e59a4c67ba3a92f7992587ac7a0408c2/project/github/RedSparr0w/node-csgo-parser/master.json'
+    '/build/gh/RedSparr0w/node-csgo-parser/master.json?token=b90b5c49e59a4c67ba3a92f7992587ac7a0408c2'
   )
   .expectBadge({
     label: 'build',
     message: isBuildStatus,
   })
 
-t.create('circle ci (not found)')
-  .get('/project/github/PyvesB/EmptyRepo.json')
-  .expectBadge({ label: 'build', message: 'project not found' })
-
-t.create('circle ci (no response data)')
+t.create('legacy route with VCS')
   .get('/project/github/RedSparr0w/node-csgo-parser.json')
-  .intercept(nock =>
-    nock('https://circleci.com')
-      .get(
-        '/api/v1.1/project/github/RedSparr0w/node-csgo-parser?filter=completed&limit=1'
-      )
-      .reply(200)
-  )
-  .expectBadge({ label: 'build', message: 'unparseable json response' })
+  .expectBadge({
+    label: 'build',
+    message: isBuildStatus,
+  })
 
-// we're passing &limit=1 so we expect exactly one array element
-t.create('circle ci (invalid json)')
-  .get('/project/github/RedSparr0w/node-csgo-parser.json')
-  .intercept(nock =>
-    nock('https://circleci.com')
-      .get(
-        '/api/v1.1/project/github/RedSparr0w/node-csgo-parser?filter=completed&limit=1'
-      )
-      .reply(200, [{ status: 'success' }, { status: 'fixed' }])
+t.create('legacy route (assume "github" as a default VCS)')
+  .get('/project/RedSparr0w/node-csgo-parser/master.json')
+  .expectBadge({
+    label: 'build',
+    message: isBuildStatus,
+  })
+
+t.create('legacy route with token')
+  .get(
+    '/token/b90b5c49e59a4c67ba3a92f7992587ac7a0408c2/project/github/RedSparr0w/node-csgo-parser/master.json'
   )
   .expectBadge({
     label: 'build',
-    message: 'invalid response data',
-    color: 'lightgrey',
+    message: isBuildStatus,
   })
```

Comment:
We admittedly don't have anything documented around our practices for redirectors yet (I opened #3402 to track creating those docs), but so far we've been testing them by disabling `followRedirect` and then checking the status and target url, example below from the vso redirector:

```js
t.create('Build: default branch')
  .get('/build/totodem/8cf3ec0e-d0c2-4fcd-8206-ad204f254a96/2.svg', {
    followRedirect: false,
  })
  .expectStatus(301)
  .expectHeader(
    'Location',
    '/azure-devops/build/totodem/8cf3ec0e-d0c2-4fcd-8206-ad204f254a96/2.svg'
  )
```

The approach you've taken here obviously validates the redirect route as well. This isn't a blocker for this PR by any means, but I just wanted to make note of it. Happy to shift this topic over to #3402 for discussion in the event folks have opinions on whether we have any preference for redirector testers (leave it up to the developer, prefer one or the other, etc.)

Reply:
Yeah I can see reasons to go both ways on this.
On the one hand, the thing we're testing is a redirector, so checking it issues a 301 is a unit testing approach to making sure that component does exactly what its supposed to do.
On the other hand, our service tests are really high-level integration tests, so testing the end-to-end behaviour also makes sense in this context.
There probably isn't a single correct answer to that, but it does make sense to pick one approach and document it as a convention. I'd be happy to go either way on it tbh.

Reply:
Agreed there are reasons to go both ways here. When I wrote the first of these I picked this way, though to be honest neither way seems perfect.

I'm inclined to do what's easy to maintain.

A sort of middle-ground option is adding a helper that does a bit of both: check the redirect, then follow it, and finally do a basic check that the final result is not an error badge. Part unit test, part integration test, with a very lightweight assertion.